website/blog/2023-10-02-reduce-agg.mdx

```diff
@@ -0,0 +1,178 @@
+
+```
+auto s = initialState;
+for (auto x : input) {
+   s = inputFunction(s, x);
+}
+
+return s;
+```
+
+Final aggregation processes a list of intermediate states and computes the final state.
+
+```
+auto s = intermediates[0];
+for (auto i = 1; i < intermediates.size(); ++i)
+   s = combineFunction(s, intermediates[i]);
+}
+
+return s;
+```
+
+For example, one can implement SUM aggregation using reduce_agg as follows:
+
+```
+reduce_agg(c, 0, (s, x) -> s + x, (s, s2) -> s + s2)
+```
+
+Implementation of AVG aggregation is a bit trickier. For AVG, state is a tuple of sum and
+count. Hence, reduce_agg can be used to compute (sum, count) pair, but it cannot compute
+the actual average. One needs to apply a scalar function on top of reduce_agg to get the
+average.
+
+```
+SELECT id, sum_and_count.sum / sum_and_count.count FROM (
+  SELECT id, reduce_agg(value, CAST(row(0, 0) AS row(sum double, count bigint)),
+    (s, x) -> CAST(row(s.sum + x, s.count + 1) AS row(sum double, count bigint)),
+    (s, s2) -> CAST(row(s.sum + s2.sum, s.count + s2.count) AS row(sum double, count bigint))) AS sum_and_count
+  FROM t
+  GROUP BY id
+);
+```
+
+The examples of using reduce_agg to compute SUM and AVG are for illustrative purposes.
+One should not use reduce_agg if a specialized aggregation function is available.
+
+One use case for reduce_agg we see in production is to compute a product of input values.
+
+```
+reduce_agg(c, 0, (s, x) -> s * x, (s, s2) -> s * s2)
+```
+
+Another example is to compute a list of top N distinct values from all input arrays.
+
+```
+reduce_agg(x, array[],
+            (a, b) -> slice(reverse(array_sort(array_distinct(concat(a, b)))), 1, 1000),
+            (a, b) -> slice(reverse(array_sort(array_distinct(concat(a, b)))), 1, 1000))
+```
+
+Note that this is equivalent to the following query:
+
+```
+SELECT array_agg(v) FROM (
+    SELECT DISTINCT v
+    FROM t, UNNEST(x) AS u(v)
+    ORDER BY v DESC
+    LIMIT 1000
+)
+```
+
+## Implementation
+
+Efficient implementation of reduce_agg lambda function is not straightforward. Let’s
+consider the logic for partial aggregation.
+
+```
+auto s = initialState;
+for (auto x : input) {
+   s = inputFunction(s, x);
+}
+```
+
+This is a data-dependent loop, i.e. the next loop iteration depends on the results of
+the previous iteration. inputFunction needs to be invoked on each input value `x`
+separately. Since inputFunction is a user-defined lambda, invoking inputFunction means
+evaluating an expression. And since expression evaluation in Velox is optimized for
+processing large batches of values at a time, evaluating expressions on one value at
+a time is very inefficient. To optimize the implementation of reduce_agg we need to
+reduce the number of times we evaluate user-defined lambdas and increase the number
+of values we process each time.
+
+One approach is to
+
+1. convert all input values into states by evaluating inputFunction(initialState, x);
```

Comment:
```
1.
2.
3.
```
?

Reply:
The syntax for numbered lists is "number followed by a period", but only first number matters. The rest of the numbers will be auto-generated as a sequence starting from the first number.